lib/coll.gi

```diff
@@ -3049,6 +3053,20 @@ end);
 InstallMethod( CanComputeIsSubset,"default: no, unless identical",
   [IsObject,IsObject],IsIdenticalObj);
 
+# avoid immediate methods triggering multiple type changes once the Size of
+# an object is known.
+InstallOtherMethod(SetSize,true,[IsObject and IsAttributeStoringRep,IsObject],
+  100, # override system setter
+function(obj,sz)
+  if not HasSize(obj) then
+    SetIsEmpty(obj,sz=0);
+    SetIsTrivial(obj,sz=1);
+    SetIsNonTrivial(obj,sz<>1);
+    SetIsFinite(obj,sz<>infinity);
```

Comment:
Doesn't this still change the type four times (less for sz=0 and sz=1 because of implications, but stilll).  Plus a fifth time for the size. So, how about something like this:
```gap
function(obj, size)
local filt;
  if HasSize(obj) and Size(obj) <> size then
    ErrorNoReturn("size of <obj>" already set to ", Size(obj), ", cannot change it to ", size);
  fi;
  if size = 0 then filt := IsEmpty;
  elif size = 1 then filt := IsTrivial;
  elif size = infinity then filt := IsNonTrivial and HasIsFinite;
  else filt := IsNonTrivial and IsFinite;
  fi;
  filt := filt and HasSize;
  obj!.Size := size;
  SetFilterObj(obj, filt);
end;
```

This assumes we already have or add (negative) implications between IsEmpty, IsTrivial and IsNonTrivial (the three are after all mutually exclusive).

(We may also wish to add `IsNonEmpty` and `IsInfinite`, but better not in this PR)

Reply:
This is done. However (the manual specifies that an `SetSize` if size is given has no effect) I have wrapped the error into a check that requires assertion level to be >=3.